Exports add fieldsdd fields to `azurerm_*_cost_management_export`: `partition_data` and `export_data_options.data_version`

joshk0 opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review


## Description

Add fields to `azurerm_*_cost_management_export`: `partition_data` and `export_data_options.version`.
    
`export_data_options.version` maps to `export.definition.dataset.configuration.dataVersion` - No default. Versions can proliferate and even though we can collect a large list of possible versions it's better to leave it free form.
    
`partition_data` maps to `export.partition_data`. Defaults to false.

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_*_cost_management_export` - support for the `partition_data` and `export_data_options.data_version` properties [GH-23747]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #23747 


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

joshk0 commented:
Please let me know if I need to run the live Azure acceptance tests myself in order to proceed.

joshk0 commented:
Hello, is it possible to get a review for this?

katbyte commented on website/docs/r/billing_account_cost_management_export.html.markdown:
```diff
@@ -60,6 +62,8 @@ The following arguments are supported:
 
 * `billing_account_id` - (Required) The id of the billing account on which to create an export. Changing this forces a new resource to be created.
 
+* `partition_data` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```
```suggestion
* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```
or potentially
```suggestion
* `partitioned_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```

katbyte reviewed (changes requested):
thanks for the pr @joshk0 - for booleans we add _enabled to the end of them to mark them as bools. once that change is made i think this is good to go

joshk0 commented:
@katbyte, done. Please have a look. Thank you!

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -29,6 +29,14 @@ func (br costManagementExportBaseResource) arguments(fields map[string]*pluginsd
+		"partition_data_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Required: false,
```
We don't need to specify `false` for `Required` since it's implied by `Optional` being set to `true`
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -100,10 +108,17 @@ func (br costManagementExportBaseResource) arguments(fields map[string]*pluginsd
 							string(exports.TimeframeTypeTheLastMonth),
 							string(exports.TimeframeTypeWeekToDate),
 							string(exports.TimeframeTypeMonthToDate),
+							string(exports.TimeframeTypeMonthToDate),
```
We already support this value above so this would be a duplicate
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -100,10 +108,17 @@ func (br costManagementExportBaseResource) arguments(fields map[string]*pluginsd
+					"data_version": {
+						Type:     pluginsdk.TypeString,
+						Optional: true,
+						Required: false,
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -100,10 +108,17 @@ func (br costManagementExportBaseResource) arguments(fields map[string]*pluginsd
 						}, false),
 					},
+
+					"data_version": {
```
Is there some validation we could add for this property? Even just a `validation.StringIsNotEmpty`?

joshk0 replied:
Yeah, that's the best we can do unfortunately. It's usually of the form yyyy-mm-dd, but then with some optional -preview or -internal suffixes. It doesn't feel right to constrain to those values since they don't seem structured other than "might contain a date"

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -174,6 +189,10 @@ func (br costManagementExportBaseResource) readFunc(scopeFieldName string) sdk.R
 
 			if model := resp.Model; model != nil {
 				if props := model.Properties; props != nil {
+					if partitionData := props.PartitionData; partitionData != nil {
+						metadata.ResourceData.Set("partition_data_enabled", *partitionData)
+					}
```
We could use `pointer.From` here to do the nil-check

```suggestion
					metadata.ResourceData.Set("partition_data_enabled", pointer.From(props.PartitionData))
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -272,6 +291,7 @@ func createOrUpdateCostManagementExport(ctx context.Context, client *exports.Exp
 	props := exports.Export{
 		ETag: etag,
 		Properties: &exports.ExportProperties{
+			PartitionData: pointer.FromBool(metadata.ResourceData.Get("partition_data_enabled").(bool)),
```
```suggestion
			PartitionData: pointer.To(metadata.ResourceData.Get("partition_data_enabled").(bool)),
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -321,7 +341,18 @@ func expandExportDefinition(input []interface{}) *exports.ExportDefinition {
+	if v, ok := attrs["data_version"]; ok {
+		dataset = &exports.ExportDataset{
+			Configuration: &exports.ExportDatasetConfiguration{
+				DataVersion: utils.String(v.(string)),
```
The `utils` package is being deprecated, we should use `pointer.To` here
```suggestion
				DataVersion: pointer.To(v.(string)),
```

stephybun commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -373,10 +404,16 @@ func flattenExportDefinition(input *exports.ExportDefinition) []interface{} {
 	}
 
+	version := ""
+	if input.DataSet != nil && input.DataSet.Configuration != nil && input.DataSet.Configuration.DataVersion != nil {
+		version = *input.DataSet.Configuration.DataVersion
```
Can make use of `pointer.From` here again 
```suggestion
	if input.DataSet != nil && input.DataSet.Configuration != nil {
		version = pointer.From(input.DataSet.Configuration.DataVersion)
```

stephybun commented on internal/services/costmanagement/export_resource_group_resource_test.go:
```diff
@@ -127,14 +127,16 @@ resource "azurerm_resource_group_cost_management_export" "test" {
   recurrence_type              = "Monthly"
   recurrence_period_start_date = "%sT00:00:00Z"
   recurrence_period_end_date   = "%sT00:00:00Z"
+  partition_data_enabled               = true
```
The basic test should only include properties that are `Required`, since this is an `Optional` property this shouldn't be here 
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_group_resource_test.go:
```diff
@@ -127,14 +127,16 @@ resource "azurerm_resource_group_cost_management_export" "test" {
   export_data_options {
-    type       = "Usage"
-    time_frame = "TheLastMonth"
+    data_version = "2019-11-01"
```
Same for this property
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_group_resource_test.go:
```diff
@@ -195,15 +199,17 @@ resource "azurerm_resource_group_cost_management_export" "import" {
   recurrence_type              = azurerm_resource_group_cost_management_export.test.recurrence_type
   recurrence_period_start_date = azurerm_resource_group_cost_management_export.test.recurrence_period_start_date
   recurrence_period_end_date   = azurerm_resource_group_cost_management_export.test.recurrence_period_start_date
+  partition_data_enabled               = azurerm_resource_group_cost_management_export.test.partition_data
```
We only specify what was supplied in the configuration used to create the resource, since we're using the basic config which should not include `Optional` properties, this should be removed 
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_group_resource_test.go:
```diff
@@ -195,15 +199,17 @@ resource "azurerm_resource_group_cost_management_export" "import" {
   export_data_options {
-    type       = "Usage"
-    time_frame = "TheLastMonth"
+    data_version = "2019-11-01"
```
Same for this property 
```suggestion
```

stephybun commented on internal/services/costmanagement/export_subscription_resource_test.go:
```diff
@@ -132,15 +132,17 @@ resource "azurerm_subscription_cost_management_export" "test" {
   recurrence_type              = "Monthly"
   recurrence_period_start_date = "%sT00:00:00Z"
   recurrence_period_end_date   = "%sT00:00:00Z"
+  partition_data_enabled               = true
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_subscription_resource_test.go:
```diff
@@ -132,15 +132,17 @@ resource "azurerm_subscription_cost_management_export" "test" {
   export_data_options {
-    type       = "Usage"
-    time_frame = "TheLastMonth"
+    data_version    = "2019-11-01"
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_subscription_resource_test.go:
```diff
@@ -207,15 +210,17 @@ resource "azurerm_subscription_cost_management_export" "import" {
   recurrence_type              = azurerm_subscription_cost_management_export.test.recurrence_type
   recurrence_period_start_date = azurerm_subscription_cost_management_export.test.recurrence_period_start_date
   recurrence_period_end_date   = azurerm_subscription_cost_management_export.test.recurrence_period_start_date
+  partition_data_enabled               = azurerm_subscription_cost_management_export.test.partition_data
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_subscription_resource_test.go:
```diff
@@ -207,15 +210,17 @@ resource "azurerm_subscription_cost_management_export" "import" {
   export_data_options {
-    type       = "Usage"
-    time_frame = "TheLastMonth"
+    data_version = "2019-11-01"
```
```suggestion
```

stephybun commented on website/docs/r/billing_account_cost_management_export.html.markdown:
```diff
@@ -60,6 +62,8 @@ The following arguments are supported:
 
 * `billing_account_id` - (Required) The id of the billing account on which to create an export. Changing this forces a new resource to be created.
 
+* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```
```suggestion
* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption. Changing this forces a new resource to be created. Defaults to `false`.
```

stephybun commented on website/docs/r/resource_group_cost_management_export.html.markdown:
```diff
@@ -59,6 +61,8 @@ The following arguments are supported:
 
 * `resource_group_id` - (Required) The id of the resource group on which to create an export. Changing this forces a new resource to be created.
 
+* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```
```suggestion
* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption. Changing this forces a new resource to be created. Defaults to `false`.
```

stephybun commented on website/docs/r/subscription_cost_management_export.html.markdown:
```diff
@@ -61,6 +63,8 @@ The following arguments are supported:
 
 * `subscription_id` - (Required) The id of the subscription on which to create an export. Changing this forces a new resource to be created.
 
+* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption.
```
```suggestion
* `partition_data_enabled` - (Optional) Whether data will be partitioned into smaller files to facilitate consumption. Changing this forces a new resource to be created. Defaults to `false`.
```

stephybun commented on internal/services/costmanagement/export_billing_account_resource_test.go:
```diff
@@ -143,15 +143,17 @@ resource "azurerm_billing_account_cost_management_export" "test" {
   recurrence_type              = "Monthly"
   recurrence_period_start_date = "%sT00:00:00Z"
   recurrence_period_end_date   = "%sT00:00:00Z"
+  partition_data_enabled               = true
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_billing_account_resource_test.go:
```diff
@@ -143,15 +143,17 @@ resource "azurerm_billing_account_cost_management_export" "test" {
   export_data_options {
-    type       = "Usage"
-    time_frame = "TheLastMonth"
+    data_version = "2019-11-01"
```
```suggestion
```

stephybun commented on internal/services/costmanagement/export_resource_group_resource_test.go:
```diff
@@ -171,14 +173,16 @@ resource "azurerm_resource_group_cost_management_export" "test" {
   recurrence_type              = "Monthly"
   recurrence_period_start_date = "%sT00:00:00Z"
   recurrence_period_end_date   = "%sT00:00:00Z"
+  partition_data_enabled               = true
```
This would trigger the recreation of the resource so this needs to be removed from the `update` test config.
```suggestion
```

stephybun reviewed (changes requested):
Thanks for this PR @joshk0. There are some minor things that need fixing up to bring this more in-line with other resources in the provider. 

Most resources have a `basic` and a `complete` test configuration, where the `basic` only includes properties that are `Required` and the `complete` configuration includes properties that are `Optional`. I see that we're missing a `complete` test config here, do you think you could look into adding that so we can validate the `partition_data_enabled` property?

joshk0 commented:
Hey there,

> Most resources have a `basic` and a `complete` test configuration, where the `basic` only includes properties that are `Required` and the `complete` configuration includes properties that are `Optional`. I see that we're missing a `complete` test config here, do you think you could look into adding that so we can validate the `partition_data_enabled` property?

In other words, does this mean that the tests that already exist should in fact be left totally unmodified (currently we add `data_version` and `partition_data_enabled` values in the update flow, after merging your suggestions), and we add a new `_complete` test which does set those fields?

Or is it good to test the optional fields functioning properly in the update flow?

joshk0 commented:
Sorry, both update and requiresImport test case mutate the new optional fields.

joshk0 commented:
> Thanks for making those changes @joshk0. Unfortunately now none of the test configs have `partition_data_enabled` so we're unable to test this change. As mentioned in my previous comment, we cannot add this property to the `basic` configuration since it's optional, and it cannot be added to the `update` configuration since it would trigger a `ForceNew`. Can you please add a `complete` test configuration that include this property?

Sure I was just asking a question and doing things that did not require an answer to that question (which you didn't directly answer, but did indirectly through hints in your answer.) I will add the test case now.

joshk0 commented:
@stephybun , please approve the workflows. I have added `complete` test case to each one, and reverted the `update` cases to not alter the data version.

stephybun reviewed (changes requested):
Apologies @joshk0! I must have overlooked the question you left about the test when I did my re-review.

Thank you for adding the `complete` test configuration. I've kicked off the tests but we have some test failures:

```
------- Stdout: -------
=== RUN   TestAccResourceGroupCostManagementExport_complete
=== PAUSE TestAccResourceGroupCostManagementExport_complete
=== CONT  TestAccResourceGroupCostManagementExport_complete
    testcase.go:121: Step 1/2 error: Error running apply: exit status 1
        Error: creating Scoped Export (Scope: "/subscriptions/*******/resourceGroups/acctestRG-cm-240801080753627897"
        Export Name: "accrg240801080753627897"): unexpected status 400 (400 Bad Request) with error: BadRequest: Request properties validation failed: Invalid deliveryInfo destination rootFolderPath; value cannot begin with a forward slash
          with azurerm_resource_group_cost_management_export.test,
          on terraform_plugin_test.tf line 47, in resource "azurerm_resource_group_cost_management_export" "test":
          47: resource "azurerm_resource_group_cost_management_export" "test" {
        creating Scoped Export (Scope:
        "/subscriptions/*******/resourceGroups/acctestRG-cm-240801080753627897"
        Export Name: "accrg240801080753627897"): unexpected status 400 (400 Bad
        Request) with error: BadRequest: Request properties validation failed:
        Invalid deliveryInfo destination rootFolderPath; value cannot begin with a
        forward slash
--- FAIL: TestAccResourceGroupCostManagementExport_complete (132.73s)
FAIL
```

```
------- Stdout: -------
=== RUN   TestAccSubscriptionCostManagementExport_update
=== PAUSE TestAccSubscriptionCostManagementExport_update
=== CONT  TestAccSubscriptionCostManagementExport_update
    testcase.go:121: Step 1/6 error: Error running apply: exit status 1
        Error: creating Scoped Export (Scope: "/subscriptions/*******"
        Export Name: "accs240801080753607043"): unexpected status 400 (400 Bad Request) with error: BadRequest: Request properties validation failed: Export data version:  is not supported.
          with azurerm_subscription_cost_management_export.test,
          on terraform_plugin_test.tf line 52, in resource "azurerm_subscription_cost_management_export" "test":
          52: resource "azurerm_subscription_cost_management_export" "test" {
        creating Scoped Export (Scope:
        "/subscriptions/*******"
        Export Name: "accs240801080753607043"): unexpected status 400 (400 Bad
        Request) with error: BadRequest: Request properties validation failed: Export
        data version:  is not supported.
--- FAIL: TestAccSubscriptionCostManagementExport_update (157.92s)
FAIL
```

joshk0 commented:
Hi @stephybun , just pushed some fixes that I think will work? If you could kick off the flows when you have a chance.

stephybun reviewed (commented):
@joshk0 unfortunately we're still getting the same failures: 
```
=== CONT  TestAccResourceGroupCostManagementExport_basic
    testcase.go:121: Step 1/2 error: Error running apply: exit status 1
        Error: creating Scoped Export (Scope: "/subscriptions/*******/resourceGroups/acctestRG-cm-240807120359725390"
        Export Name: "accrg240807120359725390"): unexpected status 400 (400 Bad Request) with error: BadRequest: Request properties validation failed: Export data version:  is not supported.
```

joshk0 commented:
@stephybun one more time please

joshk0 commented:
I had to vendor so one more time again sorry @stephybun

stephybun reviewed (changes requested):
Still receiving the same error:
```
        Error: creating Scoped Export (Scope: "/subscriptions/*******"
        Export Name: "accs240808092720759225"): unexpected status 400 (400 Bad Request) with error: BadRequest: Request properties validation failed: Export data version:  is not supported.
```

I currently don't have the capacity to dig deeper into this, best guess based on the error message being returned by the API
- Property isn't properly supported by the API, I had a brief look in the portal and I don't see an option for data version
- Value being used for data version isn't valid/supported

joshk0 commented:
Don't worry about trying to figure it out, I will just keep banging on it here just need the PRs to run

joshk0 commented:
@stephybun 
How do I run these tests myself? I reviewed the log of the unit tests and it seems like different unit tests are failing. I think I need to see the full output of the test run to better understand what I need to do.

joshk0 commented:
BTW, this is the failure from the unit-test phase:

```
2024-08-08T09:17:26.6137210Z panic: test timed out after 30s
2024-08-08T09:17:26.6137509Z running tests:
2024-08-08T09:17:26.6137778Z    TestAccPluginSDKAndDecoder (30s)
2024-08-08T09:17:26.6138183Z    TestAccPluginSDKAndDecoderOptionalComputed (30s)
2024-08-08T09:17:26.6138723Z    TestAccPluginSDKAndDecoderOptionalComputedOverride (30s)
2024-08-08T09:17:26.6139208Z    TestAccPluginSDKAndDecoderSets (30s)
2024-08-08T09:17:26.6139560Z    TestAccPluginSDKAndEncoder (30s)
2024-08-08T09:17:26.6145361Z    TestAccPluginSDKReturnsComputedFields (30s)
```

joshk0 commented:
> Property isn't properly supported by the API, I had a brief look in the portal and I don't see an option for data version

Yes, this is something you can only do via the console today.

> Value being used for data version isn't valid/supported

Perhaps it has to be mandatory. The reality is that in the portal you get to choose among a variety of different versions based on the export you are making.

flanakin commented on internal/services/costmanagement/export_billing_account_resource_test.go:
```diff
@@ -39,6 +39,25 @@ func TestAccBillingAccountCostManagementExport_basic(t *testing.T) {
 }
 
+func TestAccBillingAccountCostManagementExport_complete(t *testing.T) {
+	if os.Getenv("ARM_BILLING_ACCOUNT") == "" {
```
nit: Subscriptions are supported as well

flanakin commented on internal/services/costmanagement/export_resource_base.go:
```diff
@@ -104,6 +111,12 @@ func (br costManagementExportBaseResource) arguments(fields map[string]*pluginsd
+
+					"data_version": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
```
I don't think this should be optional. I'm pretty sure it's required unless the API is defaulting.

flanakin commented on internal/services/costmanagement/client/client.go:
```diff
@@ -6,10 +6,10 @@ package client
 import (
 	"fmt"
 
-	"github.com/hashicorp/go-azure-sdk/resource-manager/costmanagement/2021-10-01/exports"
```
I would probably keep the old version as well. I believe there are some gaps between what's supported in each version.

@jojoh would know more.

jojohpm replied:
If we have a provision to keep 2 versions, then we could keep both. 
2021-10-01 = Exports v1. Legacy exports. 
2023-07-01-preview = Exports v2. The new one with file partitioning, compression, file format, newer datasets etc... This is going to be the default experience going forward.

jojohpm reviewed (commented):
Added the comment.

wyattfry commented:
Hello @joshk0 , could you fix the merge conflicts?
